run with context

jose-zenledger opened the pull request:
In references to https://github.com/contribsys/faktory/issues/391

added RunWithContext and added mutex to Quiet, Terminate, and setUpWorkerProcess methods
allowing restarting of manager
ignoring jetbrains ide folder

mperham commented:
I'm not sure I want to support restarting the Manager instance. Caller should create a new instance if they want to "restart" job processing. That way there is a clear lifecycle to the object. WDYT?

jose-zenledger commented:
> I'm not sure I want to support restarting the Manager instance. Caller should create a new instance if they want to "restart" job processing. That way there is a clear lifecycle to the object. WDYT?

@mperham yeah that makes sense and reduces complexity, I'll make that change. We might want to document that `Run` should not be called again after `Terminate` or `Quiet` is called

mperham commented:
Documentation is fine but return an error too. Defined semantics should be enforced.

mperham commented on manager.go:
```diff
@@ -133,12 +161,21 @@ func (mgr *Manager) Run() {
+}
+
+// Run starts processing jobs.
+// This method does not return.
```
Needs an update.

jose-zenledger commented:
@mperham let me know if it looks good and I can add some tests and also add an example to the readme if you'd like

mperham commented:
Yes please. Both would be welcome.

On Tue, Feb 1, 2022 at 11:23 jose-zenledger ***@***.***>
wrote:

> @mperham <https://github.com/mperham> let me know if it looks good and I
> can add some tests and also add an example to the readme if you'd like
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/contribsys/faktory_worker_go/pull/58#issuecomment-1027203500>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAWX3TCPAVU7MRJ5QJM3TUZAXLJANCNFSM5NHG6ZGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jose-zenledger commented:
@mperham sorry for the delay, I updated the readme and added a test case. I did notice that some of the code is hard to test without having the data source setup. We could either add a docker-compose file and run tests inside there and define those data sources in CI as well or define interfaces that we can provide mock ones for in place of the faktory package. Either way that might be outside of the scope of this PR. I can definitely help out with that if you think it would be beneficial but let me know if you want anything else in this PR or see any issues.

mperham commented:
We've got a lint failure that I'm worried about:

```
manager.go:179:28: SA1012: do not pass a nil Context, even if a function permits it; pass context.TODO if you are unsure about which Context to use (staticcheck)
	return mgr.RunWithContext(nil)
```

I think we need to take this advice to heart and refactor. I will take care of it.

mperham commented:
I've tagged v1.6.0, thank you!

jose-zenledger commented:
Ah just realized you fixed it as well
